vllm/core/block/token_ids.py

```diff
@@ -0,0 +1,244 @@
+        hashed content.
+        """
+
+        return (left.content_hash == right.content_hash and
+                left.content_offset + left.token_count == right.content_offset)
+
+    def adjusted(self, tokens_start: int,
+                 tokens_end: int) -> Optional["TokenRangeAnnotation"]:
+        """
+        Computes a new TokenRangeAnnotation that corresponds to the same
+        content in a slice of the original token IDs.
+        """
+
+        unclamped_annotation_start = self.token_index - tokens_start
+        unclamped_annotation_end = unclamped_annotation_start + self.token_count
+
+        annotation_start = max(0, unclamped_annotation_start)
+        annotation_end = min(tokens_end - tokens_start,
+                             unclamped_annotation_end)
+
+        if annotation_start >= annotation_end:
+            # There is no overlap.
+            return None
+
+        return TokenRangeAnnotation(
+            content_hash=self.content_hash,
+            content_offset=self.content_offset + annotation_start -
+            unclamped_annotation_start,
+            token_index=annotation_start,
+            token_count=annotation_end - annotation_start)
+
+
+class TokenIds:
+    token_ids: Tuple[int, ...]
+    annotations: Tuple[TokenRangeAnnotation, ...]
+
+    def __init__(self,
+                 token_ids: Iterable[int] = (),
+                 annotations: Iterable[TokenRangeAnnotation] = ()):
+        self.token_ids = tuple(token_ids)
+        self.annotations = tuple(annotations)
+
+        # Ensure that the token annotations are monotonic.
+        current_token_index = 0
+        for annotation in self.annotations:
+            if (annotation.token_index < current_token_index
+                    or annotation.token_count < 0):
+                raise ValueError("TokenRangeAnnotations must be sorted and "
+                                 "non-overlapping.")
+
+            current_token_index = annotation.end_index
+
+        if current_token_index > len(self.token_ids):
+            raise ValueError("TokenRangeAnnotations must be entirely "
+                             "contained within the token IDs.")
+
+    @staticmethod
+    def from_sequence(sequence: Sequence, offset: int) -> "TokenIds":
+        # Since the block table is append-only, the unseen token ids are the
+        # ones after the appended ones.
+        token_ids = sequence.get_token_ids()[offset:]
+
+        # Adjust any annotations for the new token ids list.
+        adjusted_annotations = (a.adjusted(offset, offset + len(token_ids))
+                                for a in sequence.token_annotations)
+        filtered_annotations = (a for a in adjusted_annotations
+                                if a is not None)
+        sorted_annotations = sorted(filtered_annotations,
+                                    key=lambda a: a.token_index)
+        return TokenIds(token_ids, sorted_annotations)
+
+    def chunks(self,
```

Comment:
1. Good suggestion -- renamed it to `to_chunks`.
2. I think it'd be tricky to cache, but I did add a fast path in the slice operation since the structure gets sliced for each decoded token and those will always be after the last annotation (if there is any).